### PR TITLE
[Standard library] Move (Contiguous)Array and Optional equality operators into methods

### PR DESCRIPTION
**Pull request by DougGregor**
We are going to do this anyway for the conditional conformance of `Array` and `Optional` to `Equatable`. Best to do it earlier.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@moiseev , I'm doing this to minimize the changes to the standard library when we do make `Array`, `Optional`, etc. `Equatable` (again)

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: stdlib/public/core/Arrays.swift.gyb [hunk 1]
```diff
@@ -2177,61 +2177,56 @@ extension _ArrayBufferProtocol {
...
+      return true
+    }
 
-%if Self == 'ArraySlice':
+  %if Self == 'ArraySlice':
 
-  var streamLHS = lhs.makeIterator()
-  var streamRHS = rhs.makeIterator()
+    var streamLHS = lhs.makeIterator()
+    var streamRHS = rhs.makeIterator()
```
Comment (natecook1000) [inline stdlib/public/core/Arrays.swift.gyb]: Anyone know why we're still using iterators to compare slices? It must be strictly more overhead than indexing. cc @airspeedswift

Comment (airspeedswift) [reply stdlib/public/core/Arrays.swift.gyb]: I have another PR that gave them custom iterators, which had a speed benefit but was causing some compiler issues. If I revisit that, iterators could be faster.

**Pull request merged**